sos/upload/profiles/redhat.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2024 Red Hat, Inc. Jose Castillo <jcastillo@redhat.com>
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+import os
+import json
+from sos.upload.profiles import Upload
+from sos.utilities import convert_bytes, TIMEOUT_DEFAULT
+from sos.policies.auth import DeviceAuthorizationClass
+from sos import _sos as _
+
+try:
+    import requests
+    REQUESTS_LOADED = True
+except ImportError:
+    REQUESTS_LOADED = False
+
+
+class RHELUpload(Upload):
+
+    client_identifier_url = "https://sso.redhat.com/auth/"\
+        "realms/redhat-external/protocol/openid-connect/auth/device"
+    token_endpoint = "https://sso.redhat.com/auth/realms/"\
+        "redhat-external/protocol/openid-connect/token"
+
+    def __init__(self, parser=None, args=None, cmdline=None):
+
+        super().__init__(parser=parser, args=args, cmdline=cmdline)
+
+    RH_API_HOST = "https://api.access.redhat.com"
+    RH_SFTP_HOST = "sftp://sftp.access.redhat.com"
+    _upload_url = RH_SFTP_HOST
+    _upload_method = 'post'
+    _device_token = None
+    _upload_directory = '/'
+    # Max size for an http single request is 1Gb
+    _max_size_request = 1073741824
+
+    def prompt_for_upload_user(self):
+        if self.commons['cmdlineopts'].upload_user:
+            self.ui_log.info(
+                _("The option --upload-user has been deprecated in favour"
+                  " of device authorization in RHEL")
+            )
+        if not self.commons['cmdlineopts'].case_id:
+            # no case id provided => failover to SFTP
+            self.upload_url = self.RH_SFTP_HOST
+            self.ui_log.info("No case id provided, uploading to SFTP")
+
+    def prompt_for_upload_password(self):
+        # With OIDC we don't ask for user/pass anymore
+        if self.commons['cmdlineopts'].upload_pass:
+            self.ui_log.info(
+                _("The option --upload-pass has been deprecated in favour"
+                  " of device authorization in RHEL")
+            )
+
+    def get_upload_url(self):
+        rh_case_api = "/support/v1/cases/%s/attachments"
+        try:
+            if self.upload_url:
+                return self.upload_url
+            if self.commons['cmdlineopts'].upload_url:
+                return self.commons['cmdlineopts'].upload_url
+            if self.commons['cmdlineopts'].upload_protocol == 'sftp':
+                return self.RH_SFTP_HOST
+            if not self.commons['cmdlineopts'].case_id:
+                return self.RH_SFTP_HOST
```

Comment:
I still haven't determined the cleanest implementation for #3891, but I am considering something like:

```suggestion
            if self.commons['cmdlineopts'].upload_protocol == 'sftp':
                return self.RH_SFTP_HOST
            if self.commons['cmdlineopts'].upload_protocol == 's3':
                bucket = self.get_upload_s3_bucket()
                prefix = self.get_upload_s3_object_prefix()
                return f"s3://{bucket}/{prefix}"
            if not self.commons['cmdlineopts'].case_id:
                return self.RH_SFTP_HOST
```

This leaves off the case id as a prefix (aka directory) within the bucket and only leaves it as part of the archive name.

I had not added it to the PR to get feedback yet. I think it is likely an acceptable compromise, at least until Red Hat decides to use an s3 bucket as an additional upload target for support cases.

Reply:
I'll test this but it looks like it will work. Having it here could help receive feedback from others as well. 
And I'll make sure I add you as co-author with all these changes.